src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -6133,10 +6135,35 @@ void StorageReplicatedMergeTree::clearOldPartsAndRemoveFromZK()
     /// Delete duplicate parts from filesystem
     if (!parts_to_delete_only_from_filesystem.empty())
     {
-        clearPartsFromFilesystem(parts_to_delete_only_from_filesystem);
-        removePartsFinally(parts_to_delete_only_from_filesystem);
+        /// It can happen that some error appear during part removal from FS.
+        /// In case of such exception we have to change state of failed parts from Deleting to Outdated.
+        /// Otherwise nobody will try to remove them again (see grabOldParts).
+        NameSet parts_failed_to_delete;
+        tryClearPartsFromFilesystem(parts_to_delete_only_from_filesystem, parts_failed_to_delete);
 
-        LOG_DEBUG(log, "Removed {} old duplicate parts", parts_to_delete_only_from_filesystem.size());
+        DataPartsVector finally_remove_parts;
+        if (!parts_failed_to_delete.empty())
+        {
+            DataPartsVector rollback_parts;
+            for (const auto & part : parts_to_delete_only_from_filesystem)
+            {
+                if (!parts_failed_to_delete.contains(part->name))
+                    finally_remove_parts.push_back(part);
+                else
+                    rollback_parts.push_back(part);
+            }
+
+            if (!rollback_parts.empty())
+                rollbackDeletingParts(rollback_parts);
+        }
+        else  /// all parts was successfully removed
+        {
+            finally_remove_parts = parts_to_delete_only_from_filesystem;
+        }
+
+        removePartsFinally(finally_remove_parts);
```

Comment:
Right now, the whole code of `clearOldPartsAndRemoveFromZK` looks over-complicated and not very solid to me.
Generally, after `grabOldParts` we can get a set of parts with `Deleting` which we must delete or roll back to `Outdated`. But it is not clear if what we do in between is exception safe. Like, here we assume that `rollbackDeletingParts` and `removePartsFinally` don't throw (which is optimistic cause e.g. the last one writes to part log).

Maybe we can always keep a `parts` vector (as a result of `grabOldParts`), and also has a vector with statuses (Deleting, DeletedFromZk, DeletedFromFS, Outdated, something else) so that in the final try/catch we can (more or less) reliably roll back statuses?

BTW, `grabOldParts` is complicated as well, and it is not (very) clear if we can get exception inside and get half-state where some parts were if fact marked as deleting. However, at least a loop with `modifyPartState` is the last one:
https://github.com/ClickHouse/ClickHouse/blob/27fd255be61b479484f3428b9912f5a436e9f481/src/Storages/MergeTree/MergeTreeData.cpp#L1647-L1651

Reply:
Discussed in TG and decide to current version. After that we can apply refactoring.